### PR TITLE
1) Add PhasedFSimParameters container that describes cirq.PhasedFSimGate angles

### PR DESCRIPTION
**Pull request by mrwojtek**
This change adds PhasedFSimParameters that is a container for all the angles of cirq.PhasedFSimGate.

It is used for calibration purposes and stores only a numerical values for the angles, Since some calibrations estimate only a subset of the parameters, all of them are optional.

### CODE REVIEW

File: cirq/google/calibration/phased_fsim.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+if TYPE_CHECKING:
+    # Workaround for mypy custom dataclasses
+    from dataclasses import dataclass as json_serializable_dataclass
+else:
+    from cirq.protocols import json_serializable_dataclass
+
+
+@json_serializable_dataclass(frozen=True)
+class PhasedFSimParameters:
```
Comment (dstrain115) [inline cirq/google/calibration/phased_fsim.py]: Add doc string.

File: cirq/protocols/json_serialization.py [hunk 2]
```diff
@@ -140,6 +140,7 @@ def two_qubit_matrix_gate(matrix):
         'PhaseFlipChannel': cirq.PhaseFlipChannel,
         'PhaseGradientGate': cirq.PhaseGradientGate,
         'PhasedFSimGate': cirq.PhasedFSimGate,
+        'PhasedFSimParameters': cirq.google.PhasedFSimParameters,
```
Comment (dstrain115) [inline cirq/protocols/json_serialization.py]: If you want, you can add this to the NOT_YET_SERIALIZED list, and I can add the json serialization if you want?

Comment (mrwojtek) [reply cirq/protocols/json_serialization.py]: There is no problem with json serialization for this dataclass as far as I know (I'll double-check this). The problems are in PR 2)

**Comment by mrwojtek** [comment]
Update:
1. Rename from PhsaedFSimParameters to PhasedFSimCharacterization. This way it becomes calibration internal dataclass which most likely should be kept here. The main difference between here and the rest of cirq is that symbols are meaningless and unitary angles might be optional, which is something meaningless for the gate notion.
2. Added complete docstrings.
3. Added tests.
4. Requires styling/formatting changes.

**Comment by mrwojtek** [comment]
I have tested the JSON serialization/deserialization, it works. Would be nice to have a tests for that in the future.

File: cirq/google/calibration/phased_fsim.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+    from cirq.protocols import json_serializable_dataclass
+
+
+@json_serializable_dataclass(frozen=True)
+class PhasedFSimCharacterization:
+    """Holder for the unitary angles of the cirq.PhasedFSimGate.
+
+    This class stores five unitary parameters (θ, ζ, χ, γ and φ) that describe the
+    cirq.PhasedFSimGate which is the most general particle conserving two-qubit gate. The unitary
+    of the underlying tate is:
```
Comment (dstrain115) [inline cirq/google/calibration/phased_fsim.py]: tate to gate

File: cirq/google/calibration/phased_fsim.py [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+        phi: φ angle in radians or None when unknown.
+    """
+    theta: Optional[float] = None
+    zeta: Optional[float] = None
+    chi: Optional[float] = None
+    gamma: Optional[float] = None
+    phi: Optional[float] = None
+
+    def asdict(self) -> Dict[str, float]:
+        """Converts parameters to a dictionary that maps angles name to values.
```
Comment (dstrain115) [inline cirq/google/calibration/phased_fsim.py]: nit: I would probably change to "angle names"

File: cirq/google/calibration/phased_fsim.py [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+        """
+        return dataclasses.asdict(self)
+
+    def all_none(self) -> bool:
+        """Checks if all the angles are None."""
```
Comment (dstrain115) [inline cirq/google/calibration/phased_fsim.py]: Maybe "Returns True if all the angles are None." would be more precise.

Same below for any_none

**Review by dstrain115 — APPROVED** [review]
I am surprised that you don't have to add a json and repr file to json_test_data for this.  I guess @json_serialized_dataclass takes care of it for you.

Some minor nits, but otherwise, looks fine to me.

**Comment by mrwojtek** [comment]
You are right with JSON serialization. It was working but the Cirq serialization tests were failing since I didn't provide the necessary golden files apparently. I removed the serialization completely from this PR, this is something to be done in a follow-up.

**Comment by mpharrigan** [comment]
Is there a way to see a clean diff of this PR? It's pulling in all the changes to Cirq `master` as well

**Comment by mpharrigan** [comment]
This PR is https://github.com/quantumlib/Cirq/pull/3646, right?

**Comment by mrwojtek** [comment]
Yes

**Pull request merged**